src/gt4py/next/common.py

```diff
@@ -117,10 +116,143 @@ def __getitem__(self, index: int | slice) -> int | UnitRange:
             else:
                 raise IndexError("UnitRange index out of range")
 
+    def __and__(self, other: UnitRange) -> UnitRange:
+        start = max(self.start, other.start)
+        stop = min(self.stop, other.stop)
+
+        # Handle the case where there is no overlap
+        if stop <= start:
+            return UnitRange(0, 0)
+
+        return UnitRange(start, stop)
+
 
 DomainT: TypeAlias = tuple[tuple[Dimension, UnitRange], ...]
 
 
+def promote_dims(*dims_list: list[Dimension]) -> list[Dimension]:
```

Comment:
rename to something generic e.g. `topological_sort` and move to gt4py._core.utils (or something)
```suggestion
T = TypeVar("T", Hashable) #???
def topological_sort(*dims_list: list[T]) -> list[T]:
```

Reply:
This function is not just topologically sorting, but also ensuring the order is unique. This is a rather special requirement and one would need to enhance the error messages again (they are user facing) for the dimension case. As long as it is not used in a more generic context it seams like a violation of KISS to generalize this.

Reply:
so where do you want to put it?

Reply:
Could we leave it in `common` for now? Placing it in `gt4py._core.utils` still results in circular imports due to importing `Dimension` for the type annotations.